### PR TITLE
fix(BBD-724): When range refinement value is full bucket, don't apply any refinement

### PR DESCRIPTION
**Pull request by zfcb**
Also need to review PR in api-javascript

https://github.com/groupby/api-javascript/pull/79

### CODE REVIEW

File: src/core/action-creator.ts [hunk 1]
```diff
@@ -102,6 +102,10 @@ export function createActions(flux: FluxCapacitor) {
                 .findIndex((index) => SearchAdapter.refinementsMatch(payload, <any>navigation.refinements[index], navigation.range ? 'Range' : 'Value')) === -1;
             },
             msg: 'refinement is already selected'
+          }, {
+            // tslint:disable-next-line max-line-length
+            func: ({ range }, state) => range ? !(Selectors.rangeHighest(state, field) === high && Selectors.rangeLowest(state, field) === valueOrLow) : true,
```
Comment (effervescentia) [inline src/core/action-creator.ts]: can make this `!!range && !(/* other logic */)`

File: src/core/action-creator.ts [hunk 2]
```diff
@@ -102,6 +102,10 @@ export function createActions(flux: FluxCapacitor) {
...
+            msg: 'range must not be full bucket'
```
Comment (effervescentia) [inline src/core/action-creator.ts]: `'range must be smaller than full range'`

File: src/core/adapters/search.ts [hunk 3]
```diff
@@ -41,6 +41,8 @@ namespace Adapter {
     more: navigation.moreRefinements,
     or: navigation.or,
     range: !!navigation.range,
+    rangeHighest: navigation.rangeHighest,
```
Comment (effervescentia) [inline src/core/adapters/search.ts]: `min`, `max`

File: src/core/selectors.ts [hunk 4]
```diff
@@ -78,6 +78,12 @@ namespace Selectors {
               ? { navigationName: nav.field, type: 'Range', high, low }
               : { navigationName: nav.field, type: 'Value', value }), [])), []);
 
+  export const rangeHighest = (state: Store.State, navigationId) =>
```
Comment (effervescentia) [inline src/core/selectors.ts]: rename to `rangeNavigationMin()` and `rangeNavigationMax()`

File: src/core/store/index.ts [hunk 5]
```diff
@@ -232,6 +232,8 @@ namespace Store {
     label: string; // post
     more?: boolean; // post
     range?: boolean; // post
+    rangeHighest?: number;
```
Comment (effervescentia) [inline src/core/store/index.ts]: `min` and `max`

File: test/unit/core/adapters/search.ts [hunk 6]
```diff
@@ -87,6 +87,8 @@ suite('Search Adapter', ({ expect, stub }) => {
         more: true,
         or: true,
         range: false,
+        rangeHighest: undefined,
```
Comment (effervescentia) [inline test/unit/core/adapters/search.ts]: `min`, `max`

File: test/unit/core/selectors.ts [hunk 7]
```diff
@@ -33,6 +33,47 @@ suite('selectors', ({ expect, stub }) => {
...
+    it('should return the rangeLowest', () => {
+      const rangeLowest = 4;
+      const id = 'hello';
+      const state: any = {
+        data: {
+          present: {
+            navigations: {
+              byId: {
+                [id]: {
+                  rangeLowest
```
Comment (effervescentia) [inline test/unit/core/selectors.ts]: `min`

File: test/unit/core/selectors.ts [hunk 8]
```diff
@@ -33,6 +33,47 @@ suite('selectors', ({ expect, stub }) => {
...
+    it('should return the rangeHighest', () => {
+      const rangeHighest = 4;
+      const id = 'hello';
+      const state: any = {
+        data: {
+          present: {
+            navigations: {
+              byId: {
+                [id]: {
+                  rangeHighest
```
Comment (effervescentia) [inline test/unit/core/selectors.ts]: `max`

File: test/unit/core/selectors.ts [hunk 9]
```diff
@@ -33,6 +33,47 @@ suite('selectors', ({ expect, stub }) => {
...
+  describe('rangeLowest()', () => {
+    it('should return the rangeLowest', () => {
+      const rangeLowest = 4;
+      const id = 'hello';
+      const state: any = {
+        data: {
+          present: {
+            navigations: {
+              byId: {
+                [id]: {
```
Comment (effervescentia) [inline test/unit/core/selectors.ts]: this structure doesn't need to be fully expanded like this, try collapsing a few layers in the middle to make it cleaner:

```js
const state: any = {
  data: {
    present: {
      navigations: {
        byId: { [id]: { rangeLowest } }
      }
    }
  }
};
```

File: src/core/action-creator.ts (hunk 1 above)
Comment (jmbtutor) [reply src/core/action-creator.ts]: I'm not sure that's equivalent. The original is `true` when `range` is falsey. I think you want `!range || !(/* other logic*/)`. Also, you might benefit from DeMorgan's law for the other logic.

File: test/unit/core/selectors.ts [hunk 10]
```diff
@@ -33,6 +33,23 @@ suite('selectors', ({ expect, stub }) => {
       expect(Selectors.isRefinementDeselected(<any>{}, 'my navigation', 4)).to.be.true;
     });
   });
+  describe('rangeNavigationMin()', () => {
+    it('should return the min', () => {
+      const min = 4;
+      const id = 'hello';
+      const state: any = { data: { present: { navigations: { byId: { [id]: { min }}}}}};
+      expect(Selectors.rangeNavigationMin(state, id)).to.be.eq(min + 0);
```
Comment (effervescentia) [inline test/unit/core/selectors.ts]: separate expectation from setup with a newline

File: test/unit/core/validators.ts [hunk 11]
```diff
@@ -327,4 +327,29 @@ suite('validators', ({ expect, spy, stub }) => {
...
+    const high = 8;
+    const range = true;
+    const navigationId = 'a';
+    const payload = {
+      navigationId,
+      range,
+      low,
+      high
+    };
+    const state: any = { a: 'b' };
```
Comment (effervescentia) [inline test/unit/core/validators.ts]: separate `describe()` setup from `it()`s with a newline

File: test/unit/core/validators.ts [hunk 12]
```diff
@@ -327,4 +327,29 @@ suite('validators', ({ expect, spy, stub }) => {
...
+    it('should be valid if it is not full range', () => {
+      stub(Selectors, 'rangeNavigationMax').returns(high);
+      stub(Selectors, 'rangeNavigationMin').returns(low + 1);
```
Comment (effervescentia) [inline test/unit/core/validators.ts]: separate setup from expectation/action

File: src/core/validators.ts [hunk 13]
```diff
@@ -43,6 +43,12 @@ export const isRefinementDeselectedByValue: Validator<Actions.Payload.Navigation
   msg: 'refinement is already selected'
 };
 
+export const isNotFullRange: Validator<Actions.Payload.Navigation.AddRefinement> = {
+  // tslint:disable-next-line max-line-length
+  func: ({ navigationId, range, low, high }, state) => !range || Selectors.rangeNavigationMax(state, navigationId) !== high || Selectors.rangeNavigationMin(state, navigationId) !== low,
```
Comment (effervescentia) [inline src/core/validators.ts]: the following is equivalent but I think is clearer about the condition
```js
... => !range || !(Selectors.rangeNavigationMax(state, navigationId) === high && Selectors.rangeNavigationMin(state, navigationId) === low)
```

File: src/core/action-creator.ts [hunk 14]
```diff
@@ -86,7 +86,8 @@ export function createActions(flux: FluxCapacitor) {
             validators.isRangeRefinement,
             validators.isValidRange,
             validators.isValueRefinement,
-            validators.isRefinementDeselectedByValue
+            validators.isRefinementDeselectedByValue,
+            validators.isNotFullRange
```
Comment (effervescentia) [inline src/core/action-creator.ts]: I'd put this next to the other range validators rather than at the bottom

**Pull request merged**